### ISSUE TITLE
Losing contact with Bokeh server causes training halt

### ISSUE THREAD

**Issue by dwf** [bartvm/blocks#496]
@jbornschein just mentioned this to me and I think it's probably not desirable behaviour.

**Comment by bartvm** [comment]
Although I agree, I think we eventually we should move away of Blocks interfacing with the Bokeh plot server directly. Once we start gracefully handling connection issues, things quickly start getting messy. What I would prefer to see is that Blocks gets support to log to a database (#476), and the plot server should query the database instead.

Partially it's just moving the problem, because now we need to make sure that Blocks handles disconnection to the database server gracefully, but I think that (a) will be easier and (b) be more future proof. It would also fix #450.

**Comment by jbornschein** [comment]
I agree -- having a log (and plot) server that just receives updates and decides what and how to plot afterwards seems better. In that case, maybe blocks should not even talk directly to the database sever, but rather to a web-service that keeps track of experiments and makes sure everything get properly stored.  I started a Django-based project called 'exp-db' some time ago -- but it never got to a state that it was actually usable. It would allow you to browse through running and finished experiments, add tags and comments and create plots from 'channels'. Maybe it would be time to revisit that idea. 

In the meantime I think it would be nice if a lost bokeh-server connection would not make your experiment crash and make you loose all your results. It is annoying because especially long-running experiments have a higher probability of loosing the connection -- those experiments where it hurts most.

**Comment by bartvm** [comment]
Yeah, a quick fix would be welcome.

I'm not sure about sending data to a web API instead of a database server directly, seems like just another indirection (and speed is quite important if you're sending log data potentially multiple times per second). My suggestion would be to dump the entire log directly in MongoDB, and then build a web interface around that database instead of the other way around.

I actually started a Django-based project for Pylearn2 as well :) but never got it into a workable state either. I very much like the idea though. My ideal would actually be to have a single MongoDB server in the lab with a web interface up and running that allows people to basically start up Blocks, which starts sending data to the database, and open up a web interface that shows near live plotting, make notes, export images, etc.

**Comment by dwf** [comment]
> single MongoDB server in the lab

ಠ_ಠ

Given the way everything else regarding lab infrastructure is run, I think
that would probably be quite undesirable. And no matter how idiot proof we
make it the world will build a better idiot.

**Comment by bartvm** [comment]
Haha, that might be true. Jobman does the same though, and can send data back from the clusters to a PostgreSQL database running on `opter`, right? Clusters are just one example of a case where you can't or don't want to be bothered setting up a locally running MongoDB instance.

For me it's a huge plus to be able to collect the results of many experiments in a central database. The current workflow, which for most people (at least myself) involves either having logs saved as part of huge models (which are often deleted or overwritten), or having separate log files (text, pickled Python objects) lying all over the place, is really quite primitive, and makes for frustrating analysis over multiple runs.

Having a single MongoDB server would then just be a way of letting everyone benefit from centralized logging, without having to walk every single person through getting their own local MongoDB instance up and running.

**Comment by rizar** [comment]
I think @dwf is right that a centralized server is utopia. A server for a project sounds more realistic.

**Comment by bartvm** [comment]
Regardless of the technical difficulties of maintaining a MongoDB server within the lab's infrastructure, the point is that the possibility is there. It's an option that could potentially scale, whereas the other options simply can't.
